drivers/soft_uart/soft_uart.c

```diff
@@ -0,0 +1,323 @@
+/*
+ * Copyright (C) 2020 ML!PA Consulting GmbH
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_soft_uart
+ * @{
+ *
+ * @file
+ * @brief       Software UART implementation
+ *
+ * @author      Benjamin Valentin <benjjamin.valentin@ml-pa.com>
+ */
+
+#include <stdio.h>
+#include <assert.h>
+
+#include "mutex.h"
+#include "soft_uart.h"
+#include "soft_uart_params.h"
+
+enum {
+    STATE_RX_IDLE,
+    STATE_RX_HIGH,
+    STATE_RX_LOW
+};
+
+enum {
+    PARITY_NONE,
+    PARITY_EVEN,
+    PARITY_ODD,
+    PARITY_MARK,
+    PARITY_SPACE,
+};
+
+struct uart_ctx {
+    mutex_t lock;       /**< UART mutex          */
+    mutex_t sync;       /**< TX byte done signal */
+    uart_rx_cb_t rx_cb; /**< RX callback         */
+    void* rx_cb_arg;    /**< RX callback arg     */
+    uint32_t baud;      /**< timer ticks per bit */
+    uint16_t byte_tx;   /**< current TX byte     */
+    uint16_t byte_rx;   /**< curretn RX byte     */
+    uint8_t bits_tx;    /**< TX bit pos          */
+    uint8_t state_rx;   /**< RX state            */
+#ifdef MODULE_SOFT_UART_MODECFG
+    uint8_t data_bits;  /**< number of data bits */
+    uint8_t stop_bits;  /**< number of stop bits */
+    uint8_t parity;     /**< parity mode         */
+#endif
+} soft_uart_ctx[SOFT_UART_NUMOF];
```

Comment:
I think the coding convention wants us to `typedef` the shit out of it...

Reply:
since this is an internal struct, do we really need a typedef for it?

Reply:
Apparently [I was wrong](https://github.com/RIOT-OS/RIOT/blob/master/CODING_CONVENTIONS.md#types): It now states that "`typedef`s are allowed", not that they are mandatory :-)

I'm personally in favor of the [Linux coding convention](https://www.kernel.org/doc/html/latest/process/coding-style.html#typedefs) regarding this issue anyway.